File: VB/WindowsApplication54/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Reflection
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports DevExpress.XtraReports.Extensions
Imports DevExpress.XtraReports.UI
' ...

Namespace WindowsApplication54
	Partial Public Class Form1
		Inherits Form
		Shared Sub New()
			ReportExtension.RegisterExtensionGlobal(New ReportExtension())
			ReportDesignExtension.RegisterExtension(New DesignExtension(), ExtensionName)
			Dim fileName As String = _ 
			    Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location) & "\person.mdb"
			If (Not File.Exists(fileName)) Then
				Dim session As UnitOfWork = CreateSessionByName("person.mdb")
				If New XPCollection(Of Person)(session).Count < 6 Then
					Person.Create(session, "Name1")
					Person.Create(session, "Name2")
					Person.Create(session, "Name3")
					Person.Create(session, "Name4")
					Person.Create(session, "Name5")
					Person.Create(session, "Name6")
					session.CommitChanges()
				End If
			End If
		End Sub
		Private Shared Function CreateSessionByName(ByVal name As String) As UnitOfWork
			Dim connectionString As String = AccessConnectionProvider.GetConnectionString(name)
			Return CreateSession(connectionString)
		End Function
		Private Shared Function CreateDataSource(ByVal session As UnitOfWork) As XPCollection(Of Person)
			Return New XPCollection(Of Person)(session)
		End Function
		Public Shared Function CreateSession(ByVal connectionString As String) As UnitOfWork
			Dim result As New UnitOfWork()
			result.ConnectionString = connectionString
			Return result
		End Function

		Private Const ExtensionName As String = "Custom"

		Public Sub New()
			InitializeComponent()
		End Sub
		Private Sub createReportWhithDataSourceButton_Click(ByVal sender As Object, ByVal e As EventArgs) _ 
		    Handles createReportWhithDataSourceButton.Click
			Using session As UnitOfWork = CreateSessionByName("person.mdb")
				Dim report As New XtraReport()
				ReportDesignExtension.AssociateReportWithExtension(report, ExtensionName)
				report.DataSource = CreateDataSource(session)
				report.ShowDesignerDialog()
			End Using
		End Sub
		Private Sub createEmptyReportButton_Click(ByVal sender As Object, ByVal e As EventArgs) _ 
		    Handles createEmptyReportButton.Click
			CType(New XtraReport(), XtraReport).ShowDesignerDialog()
		End Sub
		Private Class ReportExtension
			Inherits ReportStorageExtension
			Public Overrides Sub SetData(ByVal report As XtraReport, ByVal stream As Stream)
				report.SaveLayoutToXml(stream)
			End Sub
			Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
				report.SaveLayoutToXml(url)
			End Sub
		End Class
		Private Class DesignExtension
			Inherits ReportDesignExtension
			Protected Overrides Function CanSerialize(ByVal data As Object) As Boolean
				Return TypeOf data Is XPCollection(Of Person)
			End Function
			Protected Overrides Function SerializeData(ByVal data As Object, ByVal report As XtraReport) As String
				Dim collection As XPCollection(Of Person) = TryCast(data, XPCollection(Of Person))
				If collection IsNot Nothing Then
					Return collection.Session.ConnectionString
				End If
				Return MyBase.SerializeData(data, report)
			End Function
			Protected Overrides Function CanDeserialize(ByVal value As String, ByVal typeName As String) As Boolean
				Return GetType(XPCollection(Of Person)).FullName = typeName
			End Function
			Protected Overrides Function DeserializeData(ByVal value As String, _ 
			    ByVal typeName As String, ByVal report As XtraReport) As Object
				If GetType(XPCollection(Of Person)).FullName = typeName Then
					Dim result As New XPCollection(Of Person)()
					result.Session = Form1.CreateSession(value)
					Return result
				End If
				Return MyBase.DeserializeData(value, typeName, report)
			End Function
		End Class
	End Class
End Namespace

File: CS/WindowsApplication54/Form1.cs
using System;
using System.ComponentModel;
using System.IO;
using System.Reflection;
using System.Windows.Forms;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.XtraReports;
using DevExpress.XtraReports.Extensions;
using DevExpress.XtraReports.UI;
// ...

namespace WindowsApplication54 {
    public partial class Form1 : Form {
        static Form1() {
            // The following code is required to support serialization of multiple custom objects.
            TypeDescriptor.AddAttributes(typeof(XPCollection), new ReportAssociatedComponentAttribute());

            // The following code is required to serialize custom objects.
            ReportExtension.RegisterExtensionGlobal(new ReportExtension());
            ReportDesignExtension.RegisterExtension(new DesignExtension(), ExtensionName);

            string fileName = 
                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) + "\\person.mdb";
            if (!File.Exists(fileName)) {
                UnitOfWork session = CreateSessionByName("person.mdb");
                if (new XPCollection<Person>(session).Count < 6) {
                    Person.Create(session, "Name1");
                    Person.Create(session, "Name2");
                    Person.Create(session, "Name3");
                    Person.Create(session, "Name4");
                    Person.Create(session, "Name5");
                    Person.Create(session, "Name6");
                    session.CommitChanges();
                }
            }
        }
        static UnitOfWork CreateSessionByName(string name) {
            string connectionString = AccessConnectionProvider.GetConnectionString(name);
            return CreateSession(connectionString);
        }
        static XPCollection<Person> CreateDataSource(UnitOfWork session) {
            return new XPCollection<Person>(session);
        }
        public static UnitOfWork CreateSession(string connectionString) {
            UnitOfWork result = new UnitOfWork();
            result.ConnectionString = connectionString;
            return result;
        }

        private const string ExtensionName = "Custom";

        public Form1() {
            InitializeComponent();
        }
        private void createReportWhithDataSourceButton_Click(object sender, EventArgs e) {
            using (UnitOfWork session = CreateSessionByName("person.mdb")) {
                XtraReport report = new XtraReport();
                ReportDesignExtension.AssociateReportWithExtension(report, ExtensionName);
                report.DataSource = CreateDataSource(session);
                report.ShowDesignerDialog();
            }
        }
        private void createEmptyReportButton_Click(object sender, EventArgs e) {
            new XtraReport().ShowDesignerDialog();
        }
        class ReportExtension : ReportStorageExtension {
            public override void SetData(XtraReport report, Stream stream) {
                report.SaveLayoutToXml(stream);
            }
            public override void SetData(XtraReport report, string url) {
                report.SaveLayoutToXml(url);
            }
        }
        class DesignExtension : ReportDesignExtension {
            protected override bool CanSerialize(object data) {
                return data is XPCollection<Person>;
            }
            protected override string SerializeData(object data, XtraReport report) {
                XPCollection<Person> collection = data as XPCollection<Person>;
                if (collection != null)
                    return collection.Session.ConnectionString;
                return base.SerializeData(data, report);
            }
            protected override bool CanDeserialize(string value, string typeName) {
                return typeof(XPCollection<Person>).FullName == typeName;
            }
            protected override object DeserializeData(string value, 
                string typeName, XtraReport report) {
                if (typeof(XPCollection<Person>).FullName == typeName) {
                    XPCollection<Person> result = new XPCollection<Person>();
                    result.Session = Form1.CreateSession(value);
                    return result;
                }
                return base.DeserializeData(value, typeName, report);
            }
        }
    }
}